continuous_integration/scripts/parse_stdout.py

```diff
@@ -0,0 +1,111 @@
+"""On Windows, pytest-timeout kills off the whole test suite, leaving no junit report
+behind. Parse the stdout of pytest to generate one.
+"""
+from __future__ import annotations
+
+import re
+import sys
+from collections import Counter, defaultdict
+from collections.abc import Iterable
+from datetime import datetime
+
+OUTCOMES = {
+    "PASSED",
+    "FAILED",
+    # Test timeout. Marked as a variant of FAILED in the junit report
+    None,
+    # Setup failed or teardown failed.
+    # In the latter case, if the test also failed, show both a FAILED and an ERROR line.
+    "ERROR",
+    # @pytest.mark.skip, @pytest.mark.skipif, or raise pytest.skip()
+    "SKIPPED",
+    # Reported as a variant of SKIPPED in the junit report
+    "XFAIL",
+    # These appear respectively before and after another status. Ignore.
+    "RERUN",
+    "LEAKED",
+}
+
+
+def parse_rows(rows: Iterable[str]) -> list[tuple[str, str, set[str | None]]]:
+    match = re.compile(
+        r"(distributed/.*test.*)::([^ ]*)"
+        r"( (.*)(PASSED|FAILED|ERROR|SKIPPED|XFAIL|RERUN|LEAKED).*| )$"
+    )
+
+    out: defaultdict[tuple[str, str], set[str | None]] = defaultdict(set)
+
+    for row in rows:
+        m = match.match(row)
+        if not m:
+            continue
+
+        fname = m.group(1)
+        clsname = fname.replace("/", ".").replace(".py", "").replace("::", ".")
+
+        tname = m.group(2).strip()
+        if m.group(4) and "]" in m.group(4):
+            tname += " " + m.group(4).split("]")[0] + "]"
+
+        outcome = m.group(5)
+        assert outcome in OUTCOMES
+        if outcome not in {"RERUN", "LEAKED"}:
+            out[clsname, tname].add(outcome)
+
+    return [(clsname, tname, outcomes) for (clsname, tname), outcomes in out.items()]
+
+
+def build_xml(rows: list[tuple[str, str, set[str | None]]]) -> None:
+    cnt = Counter(outcome for _, _, outcomes in rows for outcome in outcomes)
+    timestamp = datetime.now().strftime("%Y-%m-%dT%H:%M:%S.%f")
+
+    # We could have used ElementTree but it feels like overkill here
+    print('<?xml version="1.0" encoding="utf-8"?>')
```

Comment:
printing is an interesting choice. 
I assume this was done for debugging? Why not write to a file or file-like object? This way we could write the proper file out directly and/or put sys.stdout in for debugging

Reply:
No particular reason for the choice; the support in pytest for capturing stdout is good and clean so it doesn't really make a difference.

> I am wondering how stable the xml output and stdout of pytest is but I assume it is sufficient for our purposes?

Perfectly stable.